contracts/Org.sol

```diff
@@ -54,40 +58,67 @@ contract Org is Administratable {
 
   /**
    * @notice Create Organization Claim
+   * @param  claimId UUID representing this claim
    * @param  fName First name of Administrator
    * @param  lName Last name of Administrator
    * @param  eMail Email contact for Organization Administrator.
    * @param  orgAdminWalletAddress Wallet address of Organization's Administrator.
    */
   function claimRequest(
+    string memory claimId,
     string memory fName,
     string memory lName,
     string memory eMail,
     address orgAdminWalletAddress
   ) public {
+    require(
+      pendingClaims[claimId].desiredWallet == address(0),
+      "Org: Pending Claim with Id already exists"
+    );
+
     Claim memory newClaim = Claim({
       firstName: fName,
       lastName: lName,
       eMail: eMail,
-      desiredWallet: orgAdminWalletAddress,
-      filesSubmitted: true
+      desiredWallet: orgAdminWalletAddress
     });
-    emit ClaimCreated(newClaim);
-    claims.push(newClaim);
+
+    emit ClaimCreated(claimId, newClaim);
+    pendingClaims[claimId] = newClaim;
   }
 
   /**
    * @notice Approving Organization Claim
-   * @param  index Index value of Claim.
+   * @param claimId UUID of the claim being approved
    * @param adminContractAddress Contract Address for Endaoment Admin
    */
-  function approveClaim(uint256 index, address adminContractAddress)
+  function approveClaim(string calldata claimId, address adminContractAddress)
     public
     onlyAdminOrRole(adminContractAddress, IEndaomentAdmin.Role.REVIEWER)
   {
-    Claim storage claim = claims[index];
-    emit ClaimApproved(claim);
-    orgWallet = claim.desiredWallet;
+    Claim storage claim = pendingClaims[claimId];
+    require(claim.desiredWallet != address(0), "Org: Cannot approve a claim with no wallet");
+
+    emit ClaimApproved(claimId, claim);
+
+    activeClaim = claim;
+    delete pendingClaims[claimId];
+  }
+
+  /**
+   * @notice Rejecting Organization Claim
+   * @param claimId UUID of the claim being rejected
+   * @param adminContractAddress Contract Address for Endaoment Admin
+   */
+  function rejectClaim(string calldata claimId, address adminContractAddress)
```

Comment:
Questioning the utility of `rejectClaim`. Why not just `approveClaim` the good one?

Reply:
+1 I agree here. Will reduce SLOC and confusion. Only think i can think of is if someone spams a org contract and drops a ton of claims in, would that impact contract interaction GAS prices/performance? Would want a method to clear those out if so.

Reply:
Two thoughts:

1) If someone is observing the event logs to see the progression of claims, I thought it'd be useful to see a claim explicitly rejected. Maybe not needed, though.
2) If we don't have a method like this, it's not possible to clear the storage space. It doesn't cost US any gas, but it does mean the data will live in Ethereum's state bloat for all eternity.

Perhaps this is a bit of smell that the architecture is still off— maybe we want to disallow the submission of multiple claims altogether? Have a slot for a pendingClaim, it either get's accepted (clear pending, store in active) or rejected (just clear pending). Thoughts @rheeger / @CantelopePeel ?

Reply:
Ben has a good point that that we don't really even need an array, do we? 

> Have a slot for a pendingClaim, it either get's accepted (clear pending, store in active) or rejected (just clear pending)"

This sounds like a better approach to me.

The attack here is that someone runs a bot to continually fill the one pending claim with a junk claim, so to resolve this you can have an admin-only function like `rejectClaimAndApproveNewClaim` which additionally takes the same inputs as `claimRequest`. That way in one transaction you reject the existing claim and automatically approve the new one

Reply:
Is there any other way to mitigate this potential vulnerability? I am not particularly satisfied with the `rejectClaimAndApproveNewClaim` idea, but if it is the only satisfactory way to mitigate this I will settle.

Reply:
Hmm, no other good ones that I can think of currently. I personally think `rejectClaimAndApproveNewClaim` is a pretty clean solution, because you'll be verifying the claim is correct off-chain anyway, so in this event it should be simple to submit the info yourself instead of having the Org do it.

Curious to hear what you don't like about!

Reply:
Just want to make sure I understand how `rejectClaimAnApproveNewClaim` would work... 

Under this structure, only one `pendingClaim` could be submitted at a time until a decision was made about that pending claim. However, under the attack described by @mds1, someone could file a set of bogus claims and keep filing them after review to prevent anyone else from submitting a valid claim. 

`rejectClaimAndApproveNewClaim` should require either the `ADMIN` or `REVIEWER` roles to reject the given bogus `pendingClaim` while also providing the details of a valid claim to be atomically approved inside the method.

Is that accurate @mds1 ?  

If so, I think this introduces added complexity that is maybe unnecessary. It seems to me that we should go back to being able to hold multiple `pendingClaims` at a time. I know this would mean holding bloated contract state, but it inherently thwarts the described attack. 

Thoughts?

Reply:
Matt's point about the spam claims blocking legitimate ones is a good one. Robbie's point about the atomic update adding complexity is also a good one. Thinking through it, my preference would be to leave the system as is, including the `rejectClaim` method. If you choose not to use it, that's fine, but I don't think it having it has any downsides.

Reply:
@rheeger Your description was accurate! And you are correct that the current approach does prevent that attack

I'm good with @apbendi's path forward as long as it's good with everyone else. The additional gas costs at deployment for adding one function are relatively small compared to the total deployment costs. So as long as `rejectClaim` is guarded with the proper modifier (which seems to be the case) then there's really no additional risk of including it

Reply:
I approve of @apbendi's deal. Let's keep it the way it is with the mapping.

Reply:
Left this as is 👍